Unable to setup local kind-cluster using docker powered by lima VM

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area dev-productivity
/kind bug

**What happened**:
Starting a local kind cluster with `make kind-up` fails if running docker daemon on a guest VM like with [lima](https://github.com/lima-vm/lima), because the loopback addresses are attached to the host's loopback device instead of the guest loopback where the kind containers are created.

```sh
$ make kind-up
./hack/kind-up.sh \
                --cluster-name gardener-local \
                --path-kubeconfig /Users/hoehll/workspace/Github/gardener/gardener/example/provider-local/seed-kind/base/kubeconfig \
                --path-cluster-values /Users/hoehll/workspace/Github/gardener/gardener/example/gardener-local/kind/local/values.yaml \
                --with-lpp-resize-support false \

Checking loopback device lo0...
Adding IP address 172.18.255.1 to lo0...
Executing: /usr/bin/sudo /sbin/ifconfig lo0 add 172.18.255.1

Adding IP address 172.18.255.22 to lo0...
Executing: /usr/bin/sudo /sbin/ifconfig lo0 add 172.18.255.22

Setting up loopback device lo0 completed.
Creating cluster "gardener-local" ...
 ✓ Ensuring node image (kindest/node:v1.33.4) 🖼
 ✗ Preparing nodes 📦
Deleted nodes: ["gardener-local-control-plane"]
ERROR: failed to create cluster: 
command "docker run --name gardener-local-control-plane --hostname gardener-local-control-plane --label io.x-k8s.kind.role=control-plane --privileged --security-opt seccomp=unconfined --security-opt apparmor=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro -e KIND_EXPERIMENTAL_CONTAINERD_SNAPSHOTTER --detach --tty --label io.x-k8s.kind.cluster=gardener-local --net kind --restart=on-failure:1 --init=false --cgroupns=private 
--volume=/Users/hoehll/workspace/Github/gardener/gardener/example/gardener-local/kube-apiserver:/etc/gardener-local/kube-apiserver:ro --volume=/Users/hoehll/workspace/Github/gardener/gardener/example/gardener-local/controlplane:/etc/gardener/controlplane:ro --volume=/Users/hoehll/workspace/Github/gardener/gardener/dev/local-backupbuckets:/etc/gardener/local-backupbuckets --volume=/Users/hoehll/workspace/Github/gardener/gardener/dev/local-registry:/etc/gardener/local-registry 
--publish=0.0.0.0:32379:32379/TCP --publish=172.18.255.1:443:30443/TCP --publish=172.18.255.1:8132:32132/TCP --publish=172.18.255.22:8022:30022/TCP --publish=0.0.0.0:5001:5001/TCP --publish=172.18.255.1:5353:30053/TCP --publish=127.0.0.1:52073:6443/TCP -e KUBECONFIG=/etc/kubernetes/admin.conf kindest/node:v1.33.4@sha256:25a6018e48dfcaee478f4a59af81157a437f15e6e140bf103f85a2e7cd0cbbf2" 
failed with error: exit status 125
Command Output: 0cb62b3667b756665785214b4b41117d6c8f250f2f40131dfb8910b57e079007
docker: Error response from daemon: failed to set up container networking: driver failed programming external connectivity on endpoint gardener-local-control-plane (98b89a2be03434ed12aa1b948ab74865bb58bca37942f60dee19b7b7c6c8f3a9):
failed to bind host port for 172.18.255.22:8022:172.18.0.2:30022/tcp: cannot assign requested address

Run 'docker run --help' for more information
make: *** [kind-up] Error 1
```

**What you expected to happen**:
Kind cluster is able to bind to local addresses and start the nodes.

**How to reproduce it (as minimally and precisely as possible)**:
1. Setup docker using a lima VM -> https://github.com/lima-vm/lima/blob/master/templates/docker-rootful.yaml
2. Run `make kind-up`

**Anything else we need to know?**:

**Environment**:

- Gardener version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others: MacOS

--- a/hack/kind-setup-loopback-devices.sh
+++ b/hack/kind-setup-loopback-devices.sh
@@ -75,14 +75,39 @@ if ! command -v ip &>/dev/null; then
   return
 fi
 
-LOOPBACK_DEVICE=$(ip address | grep LOOPBACK | sed "s/^[0-9]\+: //g" | awk '{print $1}' | sed "s/:$//g")
-echo "Checking loopback device ${LOOPBACK_DEVICE}..."
-for address in "${LOOPBACK_IP_ADDRESSES[@]}"; do
-  if ip address show dev ${LOOPBACK_DEVICE} | grep -q $address/; then
-    echo "IP address $address already assigned to ${LOOPBACK_DEVICE}."
-  else
-    echo "Adding IP address $address to ${LOOPBACK_DEVICE}..."
-    ${SUDO}ip address add "$address" dev "${LOOPBACK_DEVICE}"
-  fi
+# In some scenarios a developer might run the kind cluster with a docker daemon that is running inside a VM.
+#
+# Most people are using Docker Desktop which is able to mirror the IP addresses 
+# added to the loopback device on the host into it's managed VM running the docker daemon.
+#
+# However if people are not using this approach,
+# docker might not be able to start the kind containers as the IP
+# is not visible for it since it's only attached to the loopback device on the host.
+#
+# We test the loopback devices by checking on the host directly as well as checking 
+# the loopback on the docker host using an container that is running in the host network.
+
+container_ip() {
+	docker run --rm --cap-add NET_ADMIN --network=host alpine ip $@
+}
+
+for ip_func in "ip" "container_ip"; do
+  LOOPBACK_DEVICE=$(${ip_func} address | grep LOOPBACK | sed "s/^[0-9]\+: //g" | awk '{print $1}' | sed "s/:$//g")
+  echo "Checking loopback device ${LOOPBACK_DEVICE} with ${ip_func}..."
+
+  for address in "${LOOPBACK_IP_ADDRESSES[@]}"; do
+    if ${ip_func} address show dev ${LOOPBACK_DEVICE} | grep -q $address/; then
+      echo "IP address $address already assigned to ${LOOPBACK_DEVICE}."
+    else
+      echo "Adding IP address $address to ${LOOPBACK_DEVICE}..."
+      if [[ ${ip_func} == "ip" ]]; then
+        sudo_ip_func=${SUDO}${ip_func}
+      else 
+        sudo_ip_func=${ip_func}
+      fi
+      ${sudo_ip_func} address add "$address" dev "${LOOPBACK_DEVICE}"
+    fi
+  done
+
+  echo "Setting up loopback device ${LOOPBACK_DEVICE} with ${ip_func} completed."
 done
-echo "Setting up loopback device ${LOOPBACK_DEVICE} completed."

